src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs

```diff
@@ -11485,10 +11491,18 @@ private Task WriteUnterminatedSqlValue(object value, MetaType type, int actualLe
                 case TdsEnums.SQLNVARCHAR:
                 case TdsEnums.SQLNTEXT:
                 case TdsEnums.SQLXMLTYPE:
+                case TdsEnums.SQLJSON:
 
                     if (type.IsPlp)
                     {
-                        if (IsBOMNeeded(type, value))
+                        if (type.NullableType == TdsEnums.SQLJSON)
+                        {
+                            // TODO : Performance and BOM check. Saurabh 
```

Comment:
What is the work that is needed under the TODO?

Reply:
@Wraith2  
Problem: A string, small or large, is being converted to a byte array causing allocation. 
That byte array is then written to network. 

2 ways of fixing this:
1. Send this out as Plp chunks. Say some chunk, smaller than the packetsize, write the Plp chunk length, write the string bytes, rinse and repeat till all the bytes can be written.
2. Do not allocate the byte[] to send out the string. Get the length, write it, and then based on the packet size, write out the UTF8 encoded bytes of the string using the Encoding/Text buffer APIs, so that the string is directly translated to the _outBuff, instead of the intermediate allocation. 

These are the solutions in my mind. However I have not looked at the encoding / text buffer APIs in detail. 

Any insights? This is the problem I had discussed on the call with you as well.

Reply:
Definitely option 2.
Getting the length to write it and then getting the bytes again is a bit wasteful but it's better than assuming we can buffer the input string and the bytes in memory at the same time. Consider a containerized scenario which has low memory working with large strings of json.

If we're working in netcore then we probably want to try and use https://learn.microsoft.com/en-gb/dotnet/api/system.text.unicode.utf8.fromutf16?view=net-8.0 which will allow us to use a ROS<char> from the string to get incremental chunks of bytes from it and write to the packets directly.

If we're only using ns2.0 level apis so we have netfx compatibility then sadly we're stuck with a UTF8 Encoder instance https://learn.microsoft.com/en-gb/dotnet/api/system.text.encoder.getbytes?view=net-8.0 which can be a little complicated to work with correctly.

Reply:
> Definitely option 2.
> Getting the length to write it and then getting the bytes again is a bit wasteful but it's better than assuming we can buffer the input string and the bytes in memory at the same time. Consider a containerized scenario which has low memory working with large strings of json.

I'm probably missing context/TDS knowledge here, but PLP seems ideal for large string data, no? You can have a single byte[] buffer, and repeatedly use Encoder.GetBytes (or Utf8.FromUtf16) to write out a chunk of the string into the buffer, prefixing it with the chunk length and sending that chunk out, no?

PLP is one thing in TDS which (if I understand it correctly) is missing in the PG protocol: everything has to be fully length-prefixed, so when sending a long string, Npgsql has to either do a full pass just to calculate the exact byte length in advance, or continuously allocate more memory as we're encoding, which means arbitrarily-large memory requirements (not good). The ability to send a huge value in multiple chunks - each length-prefixed and without having to prefix the whole thing with a length - seems very useful here.

Reply:
Yes, PLP is really useful for string data but the packetization requirements of tds also make it a but tricky. Any UTF16 codepoint can encode to 1..3 output bytes and typically if a 3 byte output is encountered and doesn't fit entirely in the output buffer it isn't output and the function would return only complete encodings. This is a problem because TDS packets must be filled to the last byte so you need to be able to get partial utf sequences.

I was talking to @GrabYourPitchforks about this on discord yesterday because I wanted to make sure my recommendations were correct. The best idea that was suggested was to use the Convert function and pass it an output buffer specifically of (space in packet + 3) bytes, this ensures that if any multibyte sequences are present they will be fully present in the output and we can then handle taking the bytes we need and temporarily storing the remaining parts of the encoded output until the next packet.

Reply:
Thanks for the context @Wraith2, that helps! But doesn't Encoder.GetBytes() exist exactly for this kind of thing? In other words, if there are only two bytes left in the packet buffer and the next character needs 3, wouldn't it populate 2 bytes and allow you to write the 3rd byte in the next one, after having sent the packet?

Reply:
Yes. That's why I suggested using an Encoder instance. It was pointed out to me that using Convert may not need an instance though which saves an allocation.

Reply:
Makes sense, thanks. Worst case, keeping an Encoder instance per physical connection and reusing it should also be fine (as I'm assuming no need for concurrent writing).

Reply:
Looks like we have a path forward.

@apoorvdeshmukh can you open a followup issue to tackle this performance improvement and add the link in the comment. We can tackle it in the next wave of change for JSON.

Reply:
Raised issue #2732 to track this.